pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -310,10 +311,21 @@ func isInstantQuery(path string) bool {
 
 func defaultInstantQueryParamsRoundTripper(next http.RoundTripper, now func() time.Time) http.RoundTripper {
 	return RoundTripFunc(func(r *http.Request) (*http.Response, error) {
-		if isInstantQuery(r.URL.Path) && !r.URL.Query().Has("time") {
+		// Check r.Form first as that's just looking up in a map, versus r.URL.Query() which parses the RawQuery again.
+		// Most of the clients send the POST urlencoded form anyway.
+		if isInstantQuery(r.URL.Path) && !r.Form.Has("time") && !r.URL.Query().Has("time") {
 			q := r.URL.Query()
 			q.Add("time", strconv.FormatInt(time.Now().Unix(), 10))
 			r.URL.RawQuery = q.Encode()
+
+			// If form was already parsed, we need to try to parse it again,
+			// otherwise upstream won't know about the new "time" value.
```

Comment:
[nit] I would just improve the comment to clarify that `ParseSeekerBodyForm()` both parses the query string and the data in the request body (if any). The data in the body will take precedence over the query string.

Reply:
Updated the comment: https://github.com/grafana/mimir/pull/3588/commits/6448e95c3845e4b2e94106f381ebea3ad03dac4f